### DISCUSSION TITLE
Minor issue with icon states.

### DISCUSSION THREAD

**Discussion by LordShekelstein** [ParadiseSS13/Paradise#15951] [Code Help]
So, I'm working on a clown-based martial arts which involves the firing of a "honkdouken" projectile. I've gotten the spell code to work, however, `icon_state` is giving me some difficulties. 

First and foremost, here is what I'm working with. 
![image](https://user-images.githubusercontent.com/81524060/116126707-f6bec480-a694-11eb-80b1-b1155796c08b.png)

The problem I've been getting is that the spell icon button thingy appears invisible as does the projectile itself as seen here: 
![image](https://user-images.githubusercontent.com/81524060/116127054-5fa63c80-a695-11eb-8efb-f04a013a2c52.png)

Here is where I have the icons stored.
![image](https://user-images.githubusercontent.com/81524060/116127234-98461600-a695-11eb-9d77-e9986addc6af.png)

I'm not to sure why this isn't working to be quite frank. Any assistance given would be very helpful.

**Comment by farie82** [comment]
Have you set the `icon_icon` value properly for the button? Same for the projectile, ensure that the `icon` variable is set properly
Also, your `button_icon_state` doesn't have a 0 or such after it while your `icon_state` in the DMI does

**Reply by LordShekelstein** [reply]
Right, so, I've added `icon_icon` values now. But its saying that its not defined. 
![image](https://user-images.githubusercontent.com/81524060/116131826-dd207b80-a69a-11eb-831e-138ea14df273.png)

**Reply by SteelSlayer** [reply]
These are spells, so `icon_icon` doesn't exist for them. Writing `action_icon_state = "honkdouken0"` as you've done should work though.

**Reply by LordShekelstein** [reply]
I just did `var/icon_icon` and it works fine.

**Reply by SteelSlayer** [reply]
Right, but it's going to be unused anywhere except in the definition of your spell.

**Reply by LordShekelstein** [reply]
Which is odd then, because it fixed my honkdouken problem, and without them, it uh, ceases to work.

**Reply by SteelSlayer** [reply]
Yeah, what you've done there is just suppress the undefined var error, which is allowing you to compile. Try removing/commenting out your `var/icon_icon` lines, and see if it works then.

**Reply by LordShekelstein** [reply]
Right, so, that works. However, my projectile is firing the staff of nothing projectile thing. Is there any fix for that?

**Reply by SteelSlayer** [reply]
Yes I see the issue. Since you're copying from the fireball spell, what you should do instead is make the honkdouken spell a subtype of fireball spell, with `fireball_type = obj/item/projectile/magic/honkdouken` so it fires your honkdouken projectile.

**Reply by LordShekelstein** [reply]
Right, so I took a bit of a break because this was frustrating me a tad. However, I've done as you suggested and made it a subtype of fireball. However, it is now saying that fireball_type is undefined. Any fixes? 
![image](https://user-images.githubusercontent.com/81524060/116248088-828a2c80-a739-11eb-85da-668b5e433dda.png)

**Reply by SteelSlayer** [reply]
You haven't made it a subtype yet.
The path of the fireball spells is:  `/obj/effect/proc_holder/spell/targeted/click/fireball`
And so your spell should be:       `/obj/effect/proc_holder/spell/targeted/click/fireball/honkdouken`

This also means you do not need the honkdouken `cast` proc at all, since its the same as fireball, and will be inherited from it. And then of course, you can remove the `var/` from `var/fireball_type` in your spell definition.

Edit: Also I forgot you'll need to change this line from:
`var/obj/item/projectile/magic/fireball/FB = new fireball_type(user.loc)` to
`var/obj/item/projectile/magic/FB = new fireball_type(user.loc)`
Inside the fireball's `cast` proc. since your houkdouken projectile is not going to be subtype of fireball.

**Reply by LordShekelstein** [reply]
![image](https://user-images.githubusercontent.com/81524060/116277502-888d0700-a753-11eb-93de-e64b19b96856.png)
![image](https://user-images.githubusercontent.com/81524060/116277646-aeb2a700-a753-11eb-9983-d9f4fa752d3b.png)
So, I think I've done it properly. However, it still isn't showing the actual projectile. As an aside: apologies for making you type out the word "honkdouken" so much.

**Reply by SteelSlayer** [reply]
At this point I don't know whats wrong without having the code. I suggest using the debugger, and stepping through the code to see what the issue is.

**Comment by S34NW** [comment]
Didn't you and I fix this @LordShekelstein ? If I recall, you had not defined where the sprite file was for your icon, adding that (so it didn't inherit the wrong file location) fixed the issue.
